hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomMessageCallProcessor.java

```diff
@@ -167,17 +168,23 @@ private void doExecutePrecompile(
             @NonNull final MessageFrame frame,
             @NonNull final OperationTracer tracer) {
         final var gasRequirement = precompile.gasRequirement(frame.getInputData());
+        final PrecompileContractResult result;
         if (frame.getRemainingGas() < gasRequirement) {
-            doHalt(frame, INSUFFICIENT_GAS);
+            result = PrecompileContractResult.halt(Bytes.EMPTY, Optional.of(INSUFFICIENT_GAS));
         } else {
             frame.decrementRemainingGas(gasRequirement);
-            final var result = precompile.computePrecompile(frame.getInputData(), frame);
+            result = precompile.computePrecompile(frame.getInputData(), frame);
+            // NOTE: the tracePrecompileCall() callback is a no-op with the current tracer
             tracer.tracePrecompileCall(frame, gasRequirement, result.getOutput());
             if (result.isRefundGas()) {
                 frame.incrementRemainingGas(gasRequirement);
             }
-            finishPrecompileExecution(frame, result, PRECOMPILE, (ActionSidecarContentTracer) tracer);
         }
+        // We must always call tracePrecompileResult() to ensure the tracer is in a consistent
```

Comment:
Do we not have the same issue with `doExecuteSystemContract`?

Reply:
I understand the comment at L177 that currently `tracePrecompileCall` is a noop, but even so, the way that `doExecutePrecompile` and `doExecuteSystemContract` are written the calls in them to `tracePrecompileCall` happen in a different order w.r.t. the check for sufficient gas, so if it was _not_ the case it was a noop the effects would be different (called in one case, not called in the other).

Reply:
A future refactor could move tracer actions - and possibly end-of-frame cleanup - into a separate object where they could be controlled under a try-with-resources and that might make this logic - in which conditional actions flow from the multi-return `start()` method to nested conditionals in `doExecute{Precompile,SystemContract}` down to `finishPrecompileExecution` easier to understand.

Reply:
Fixed same issue there, thanks @lukelee-sl !

Reply:
Agreed @david-bakin-sl , you can imagine this making a very nice application of a `try`-with-resources. 🤔